### PR TITLE
[RFC] Deprecate `srb rbi` in favor of Tapioca

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

This PR proposes to deprecate the `srb rbi` command (and sub-commands) and encourages users to use [Tapioca](https://github.com/Shopify/tapioca) instead.

We think that Tapioca will provide a better experience than `srb rbi`:

* `srb` has caused [a lot of issues](https://github.com/sorbet/sorbet/labels/srb) in the past
* `srb` isn't used at Stripe and thus not made a priority
* Tapioca is now a mature solution used for many aspects of RBI generation at Shopify
* Tapioca already support most of the features from `srb rbi` (see table below)
* Tapioca provides [RBI generation for Rails and other DSL](https://github.com/Shopify/tapioca#generating-rbi-files-for-rails-and-other-dsls)

Here's a comparison of Tapioca and `srb rbi` commands:

| srb                          | tapioca                                           |
| :--------------------------: | :-----------------------------------------------: |
| `srb init`                   | [`tapioca init`][tapioca-init]                    |
| `srb rbi config`             | [`tapioca config`][tapioca-config]                |
| `srb rbi sorbet-typed`       | [`tapioca annotations`][tapioca-annotations]      |
| `srb rbi gems`               | [`tapioca gems`][tapioca-gems]                    |
| `srb rbi hidden-definitions` | Partially covered by [`tapioca dsl`][tapioca-dsl] |
| `srb rbi todo`               | [`tapioca todo`][tapioca-todo]                    |
| `srb rbi suggest-typed`      | Delegated to [`spoom bump`][spoom-bump]           |
| `srb rbi find-gem-rbis`      | Folded inside `tapioca gems`                      |

[tapioca-init]: https://github.com/Shopify/tapioca#getting-started
[tapioca-config]: https://github.com/Shopify/tapioca#getting-started
[tapioca-annotations]: https://github.com/Shopify/tapioca#pulling-rbi-annotations-from-remote-sources
[tapioca-gems]: https://github.com/Shopify/tapioca#generating-rbi-files-for-gems
[tapioca-dsl]: https://github.com/Shopify/tapioca#generating-rbi-files-for-rails-and-other-dsls
[tapioca-todo]: https://github.com/Shopify/tapioca#rbi-files-for-missing-constants-and-methods
[spoom-bump]: https://github.com/Shopify/spoom#change-the-sigil-used-in-files

### Changes

This PR only adds a few deprecation message and does not remove any feature.

In the wiki, we suggest to also install Tapioca when adopting Sorbet:

image

Then we warn the user that the remaining documentation is about a deprecated feature:

image

When running a `srb rbi` related command, the user now gets this warning:

~~~
$ bin/srb rbi sorbet-typed
⚠️  The `srb rbi` command is being deprecated.

This command will be removed in future versions and all RBI generation will be done through Tapioca.

Please use `tapioca` instead.

Add Tapioca to your Gemfile then run `bundle install` to install it:

```
gem "tapioca", require: false, :group => :development
```

Once Tapioca is installed, simply run `tapioca init` to initialize your project with Sorbet and generate the necessary RBI files:

```
$ bundle exec tapioca init
```

For more information, please see https://github.com/Shopify/tapioca.
~~~

What do you think?

### CODE REVIEW

File: website/docs/adopting.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ systems. We'll declare them in our Gemfile and install them with Bundler:
```
Comment (jez) [inline website/docs/adopting.md:14]: I can't comment on the above section of the diff, but:

```
There are two components to Sorbet: the command line interface and runtime
systems. We'll declare them in our Gemfile and install them with Bundler:
```

suggestion:

```
There are three components to Sorbet:

1. The static type checker, which runs as a command line executable (`srb tc`)
2. The runtime package, which provides syntax for type annotations as well as certain type-friendly data structures.
3. Tools for generating [RBI files](rbi.md), which allow Sorbet to interoperate with gems and metaprogramming.

We'll declare them in our Gemfile and install them with Bundler:
```

Comment (jez) [reply website/docs/adopting.md:14]: feel free to remix

File: website/docs/adopting.md [hunk 2]
```diff
@@ -58,6 +60,30 @@ For small projects, Sorbet can run on a single file with no additional
 information. But for projects that have multiple files and depend on other gems,
 Sorbet needs to know more information to work.
 
+> **The `srb init` command is being deprecated.**
```
Comment (jez) [inline website/docs/adopting.md:63]: Do you mind if I draft a quick blog post which we could include in this PR, and could contain a longer discussion of what would change, why the change is being made, timelines, etc.?

**Review by paracycle — COMMENTED** [review]
We probably want to change (or maybe completely remove) the post-install message from the `sorbet` gem that suggests folks run `srb init` to start working with Sorbet. It would be weird to see the command that the gem tells you to run telling you that it is deprecated.

**Comment by jez** [comment]
I would also love if you switched from saying "`srb rbi` is deprecated" to either "tapioca is recommended over srb rbi"/"srb rbi is in maintenance mode" depending on the context.

I explained what maintenance mode will mean in the blog post, and you can feel free to link to that section header anywhere it would be convenient.

File: website/blog/2022-07-28-srb-tapioca.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  type.
+
+- It's not very robust to seemingly isolated failures. If there's a problem in a
+  single file or a single gem, that can sometimes cause the whole process to
+  abort.
+
+For these reasons and many more, `srb rbi hidden-definitions` is one of the
+tools that people struggle with the most when adopting Sorbet.
+
+But its one selling point is that sometimes there's simply no other tool. It's
```
Comment (froydnj) [inline website/blog/2022-07-28-srb-tapioca.md:174]:
```suggestion
But its one selling point is that sometimes there's simply no other tool. Its
```

File: website/blog/2022-07-28-srb-tapioca.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+- Minor changes that are easy to review will likely get reviewed (and
+  automatically released), but changes that merely add new features and are hard
+  to review will likely be politely rejected.
+
+- If a future Ruby version makes a breaking change which requires substantial
+  effort to fix, we may or may not have time to invest in fixing it. For this
+  reason we recommend that you switch to Tapioca at your earliest convenience.
+
+Even with thousands of change to the Sorbet repo, the `srb rbi` command has only
```
Comment (froydnj) [inline website/blog/2022-07-28-srb-tapioca.md:195]:
```suggestion
Even with thousands of changes to the Sorbet repo, the `srb rbi` command has only
```

File: website/blog/2022-07-28-srb-tapioca.md [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  example, Tapioca allows customizing how gems are loaded before attempting to
+  generate an RBI for the gem.
+
+- Tapioca has a more predictable approach to DSLs and metaprogramming. In
+  particular, Tapioca has built in support for various Rails DSLs, something
+  which `srb rbi` does not. Tapioca also provides an easy way to define
+  [custom DSL generators](https://github.com/Shopify/tapioca#writing-custom-dsl-compilers)
+  for teaching Sorbet about DSLs specific to your project.
+
+- Tapioca will do a better job at generating actually types inside the generated
```
Comment (froydnj) [inline website/blog/2022-07-28-srb-tapioca.md:114]:
```suggestion
- Tapioca will do a better job at generating actual types inside the generated
```

File: website/blog/2022-07-28-srb-tapioca.md [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+[this page in the Tapioca docs](https://github.com/Shopify/tapioca/wiki/How-does-Tapioca-compare-to-%60srb-rbi%60%3F).
+
+The biggest feature that is _only_ available in `srb rbi` (and has no exact
+analogue in Tapioca) is `srb rbi hidden-definitions`, which brings us to our
+next question.
+
+## What's happening with `srb rbi hidden-definitions`?
+
+Tapioca has no exact replacement for `srb rbi hidden-definitions`. This is the
+main reason why we have no active plans to remove the `srb rbi` command and it's
```
Comment (froydnj) [inline website/blog/2022-07-28-srb-tapioca.md:130]:
```suggestion
main reason why we have no active plans to remove the `srb rbi` command and its
```

**Review by froydnj — COMMENTED** [review]
Minor nits on the blog post, but otherwise seems reasonable!

File: website/blog/2022-07-28-srb-tapioca.md [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+<!--truncate-->
+
+![sorbet plus tapioca](/img/sorbet-plus-tapioca.svg)
+
+To elaborate, here's what's happening:
+
+- The [`tapioca`](https://github.com/Shopify/tapioca) gem is now the recommended
+  approach to [RBI generation](/docs/rbi) in projects that use Sorbet. This
+  recommendation is reflected in the Sorbet docs. Tapioca replaces (and improves
+  on!) functionality currently provided by the `srb rbi` command. Tapioca is
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:21]: Should it be `functionalities`?

```suggestion
  on!) functionalities currently provided by the `srb rbi` command. Tapioca is
```

File: website/blog/2022-07-28-srb-tapioca.md [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+  far this year.
+
+As of today Sorbet's guide to
+[Adopting Sorbet in an Existing Codebase](/docs/adopting) has been updated to
+recommend using Tapioca. The old guide remains available in the
+[repo history](https://github.com/sorbet/sorbet/blob/834c45a98cc52c2474b75fc58d06d4526888f48f/website/docs/adopting.md)
+if you need to reference it.
+
+If you'd like to start using Tapioca in a project that already uses `srb rbi`,
+[see this migration guide](#TODO).
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:44]: You can point to https://github.com/Shopify/tapioca/wiki/Migrating-to-Tapioca

File: website/blog/2022-07-28-srb-tapioca.md [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+With the announcements out of the way, let's dive into some questions we imagine
+you might have.
+
+## What is Tapioca?
+
+[Tapioca](https://github.com/Shopify/tapioca) is a Ruby gem that provides
+command line tools for generating [RBI files](/docs/rbi). From its
+documentation, "Tapioca surfaces types and methods from many sources that Sorbet
+cannot otherwise see—such as gems, Rails and other DSLs—compiles them into RBI
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:54]:
```suggestion
cannot otherwise see — such as gems, Rails and other DSLs — compiles them into RBI
```

File: website/blog/2022-07-28-srb-tapioca.md [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+## What's different between `srb rbi` and `tapioca`?
+
+From one standpoint, they're quite similar: they both use various heuristics to
+generate RBI files before `srb tc` runs on the project. Specifically, they both
+rely on Sorbet to typecheck the project itself (Tapioca is not some sort of
+alternative type checker for Ruby).
+
+Both Tapioca and `srb rbi` generate RBIs for a project's gems and support
+various generating RBI files that teach Sorbet about various kinds of
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:94]: "support various generating RBI files" sounds weird. Should it be "support generating various RBI files"?

File: website/blog/2022-07-28-srb-tapioca.md [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+1.  Use Ruby's reflection APIs to list all the classes, methods, and constants
+    that have been defined.
+
+1.  Compute a diff between that list of definitions and the list of definitions
+    that Sorbet can see statically in the codebase (including any RBIs that
+    might have already been generated).
+
+1.  Write out an RBI containing that difference.
+
+This tool is great when it works—you don't have to think about it!—but extremely
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:147]:
```suggestion
This tool is great when it works — you don't have to think about it! — but extremely
```

File: website/blog/2022-07-28-srb-tapioca.md [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+## What does "maintenance mode" mean?
+
+Maintenance mode for `srb rbi` largely codifies what was already the case with
+the `srb` tooling:
+
+- We won't be building new features into any `srb` tool (except for `srb tc`, of
+  course, which will continue to be actively improved).
+
+- Minor changes that are easy to review will likely get reviewed (and
+  automatically released), but changes that merely add new features and are hard
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:188]:
```suggestion
  automatically released), but changes that merely add new features or are hard
```

**Review by Morriar — COMMENTED** [review]
A few suggestions but this blogpost is amazing ❤️

File: website/docs/rbi.md [hunk 13]
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
...
+❯ bin/tapioca dsl <ConstantName>
+```
+
+You can read about all the DSL RBI compilers supplied by Tapioca in the
+[Tapioca's manual](https://github.com/Shopify/tapioca/blob/main/manual/compilers.md).
 
 ## The Hidden Definition RBI
 
+> **The `srb rbi hidden-definitions` is in
+> [maintenance mode](../blog/2022/07/28/srb-tapioca), please use `tapioca`
```
Comment (Morriar) [inline website/docs/rbi.md:208]: Is there a better way to link to blog posts?

**Comment by Morriar** [comment]
Update:

1. I removed the post install message
2. I updated the whole documentation to replace usages of `srb init` and friends by `tapioca`
3. I kept the `hidden-definitions` related section but moved the deprecation message there
4. I updated the deprecation message to use "maintenance mode" rather than "deprecated" and linked to the blog post
5. I added a section about `DSL RBIs`

There still are a few mentions of the `hidden-definitions` file but I don't want to remove them yet accordingly to @jez blogpost. 

I also didn't update the tests yet until we're all agreeing on the wording of the deprecation message.

**Comment by jez** [comment]
@Morriar did you force push away the blog post? I don't see it anymore

File: website/docs/rbi.md (hunk 13 above)
Comment (jez) [reply website/docs/rbi.md:208]: I think this is likely best.

File: gems/sorbet/bin/srb-rbi [hunk 14]
```diff
@@ -37,6 +37,25 @@ module Sorbet::Private::Main
       "
   end
 
+  def self.deprecation
+    <<~MESSAGE
+      #{emojify("⚠️ ", yellow("The"))} #{cyan("srb rbi")} #{yellow("command is in maintenance mode, please use")} #{cyan("Tapioca")} #{yellow("instead.")}
+
+      #{yellow("The command")} #{cyan("srb rbi")} #{yellow("will be removed in future versions and all RBI generation should be done through Tapioca.")}
+
+      Add Tapioca to your Gemfile then run #{cyan("bundle install")} to install it:
```
Comment (jez) [inline gems/sorbet/bin/srb-rbi:46]:
```suggestion
      To switch, add Tapioca to your Gemfile then run #{cyan("bundle install")} to install it:
```

File: website/blog/2022-07-28-srb-tapioca.md (hunk 7 above)
Comment (jez) [reply website/blog/2022-07-28-srb-tapioca.md:21]: I think both can be correct, I slightly prefer the way that "functionality" reads, as it presents the tool as a holistic approach, rather than a collection of small functionalities.

File: website/blog/2022-07-28-srb-tapioca.md (hunk 9 above)
Comment (jez) [reply website/blog/2022-07-28-srb-tapioca.md:54]: It doesn't render well, but these are em dashes:

image

which I prefer to use to typeset dashed, parenthetical expressions like this.

File: website/blog/2022-07-28-srb-tapioca.md (hunk 10 above)
Comment (jez) [reply website/blog/2022-07-28-srb-tapioca.md]: good catch thanks.

File: website/blog/2022-07-28-srb-tapioca.md (hunk 3 above)
Comment (jez) [reply website/blog/2022-07-28-srb-tapioca.md]: :shame:

File: website/blog/2022-07-28-srb-tapioca.md (hunk 11 above)
Comment (jez) [reply website/blog/2022-07-28-srb-tapioca.md:147]: ditto em dash comment above

File: website/docs/rbi.md [hunk 15]
```diff
@@ -28,19 +28,17 @@ We'll cover these questions in order.
...
-Note that `srb rbi update` is an alias for `srb init`.
+| I want to:                                                                 | so I'll run:                 |
+| -------------------------------------------------------------------------- | ---------------------------- |
+| Initialize a new Sorbet project, including all RBI files                   | `tapioca init`               |
+| &nbsp;                                                                     | &nbsp;                       |
+| Fetch pre-written RBIs from [`rbi-central`]                                | `tapioca rbi annotations`    |
+| (Re)generate RBIs for all gems using runtime reflection                    | `tapioca rbi gems`           |
+| (Re)generate RBIs for all DSLs such as Rails using runtime reflection      | `tapioca rbi dsl`            |
+| (Re)generate an RBI for all "hidden definitions" in a project (deprecated) | `srb rbi hidden-definitions` |
+| (Re)generate a the TODO RBI file (for missing constants)                   | `tapioca rbi todo`           |
```
Comment (jez) [inline website/docs/rbi.md:39]: Is `tapioca rbi` actually a command?

File: website/docs/rbi.md [hunk 16]
```diff
@@ -126,14 +127,17 @@ The most common way to get RBI files for gems is to generate (or regenerate)
 them. To regenerate RBI files in an already-initialized project:
 
 ```bash
-# Update only the autogenerated gem RBIs:
-srb rbi gems
+# Update only the autogenerated gem RBIs for new gems:
+❯ bin/tapioca gems
```
Comment (jez) [inline website/docs/rbi.md:131]: What is `bin/tapioca`? Should this be `bundle exec tapioca`?

File: website/blog/2022-07-28-srb-tapioca.md [hunk 17]
```diff
@@ -116,6 +116,13 @@ Tapioca has some unique benefits compared to `srb rbi`:
   method stubs. It will also generate `T.let` annotations for constants defined
   by the gem.
 
+- Tapioca pulls community-driven type annotations for public gems from
+  [rbi-central](https://github.com/Shopify/rbi-central) instead of from
+  [sorbet-typed](https://github.com/sorbet/sorbet-typed), which `srb rbi` uses.
+  The difference between the two repos is largely just about how metadata is
+  stored. It should be painless to contribute any relevant `sorbet-typed` RBI
+  files to `rbi-central` if your codebase depended on them.
```
Comment (Morriar) [inline website/blog/2022-07-28-srb-tapioca.md:124]: There is also a difference on how we check the RBIs for correctness:

* Runtime checks, we ensure that every constant/method defined in the RBI actually exist at runtime
* Static checks: We ensure that 1) the definition is not redundant with what Tapioca produces (like adding an empty scope already generated by Tapioca, or a method without a signature) 2) doesn't conflict with what Tapioca generates (like changing the arity of a method) 3) that the generated RBI + the written one do not produce any error when running Sorbet on it.

File: website/docs/rbi.md [hunk 18]
```diff
@@ -126,14 +127,17 @@ The most common way to get RBI files for gems is to generate (or regenerate)
...
+
+# Update autogenerated gem RBIs for all gems:
+❯ bin/tapioca gems --all
 ```
 
-The way this works is that `srb` will load a project's Gemfile and `require`
+The way this works is that `tapioca` will load a project's Gemfile and `require`
 each gem declared within. While loading each gem, it uses runtime reflection to
 learn what things that gem defines, and then serializes this information into a
 single RBI file for that gem.
```
Comment (jez) [inline website/docs/rbi.md:143]: Can we add a note here to link to the tapioca docs about how to further customize the `require` process?

File: website/docs/rbi.md [hunk 19]
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
 
 Hand-written RBIs for gems can come from either:
 
-- [`sorbet-typed`], a central repository for sharing hand-written RBI files with
+- [`rbi-central`], a central repository for sharing hand-written RBI files with
```
Comment (jez) [inline website/docs/rbi.md:158]: Should we add a note here explaining rbi-central vs sorbet-typed?

File: website/docs/rbi.md (hunk 16 above)
Comment (Morriar) [reply website/docs/rbi.md:131]: When you run `tapioca init`, Tapioca will create a [binstub](https://bundler.io/man/bundle-binstubs.1.html) for itself so you can run it through `bin/tapioca`

Comment (jez) [reply website/docs/rbi.md:131]: Great thanks, do you mind mentioning that `tapioca init` will create that in the `adopting.md` page?

File: website/docs/rbi.md (hunk 15 above)
Comment (Morriar) [reply website/docs/rbi.md]: Good catch, thanks!

File: website/docs/rbi.md (hunk 19 above)
Comment (Morriar) [reply website/docs/rbi.md:158]: Do you mean a deprecation warning or a link to this note in your blogpost?

File: website/blog/2022-07-28-srb-tapioca.md (hunk 17 above)
Comment (jez) [reply website/blog/2022-07-28-srb-tapioca.md:124]: I was trying to figure out how much to say "yeah it's just better" + link to the full docs, versus how much to explain in detail about why it's better.

Feel free to push whatever changes you'd like to this section.

File: website/docs/rbi.md (hunk 19 above)
Comment (jez) [reply website/docs/rbi.md:158]: I was thinking maybe just put something here like

```markdown
> **Note**: With [`srb rbi` moving to maintenance mode](blog-post), [`rbi-central`] has shifted to being the main repository of community-sourced RBI files for open source gems, replacing [`sorbet-typed`].
```

File: website/docs/rbi.md [hunk 20]
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
...
+# Update the autogenerated DSL RBI for a specific constant:
+❯ bin/tapioca dsl <ConstantName>
+```
+
+You can read about all the DSL RBI compilers supplied by Tapioca in the
+[Tapioca's manual](https://github.com/Shopify/tapioca/blob/main/manual/compilers.md).
 
 ## The Hidden Definition RBI
 
+> **The `srb rbi hidden-definitions` is in
```
Comment (jez) [inline website/docs/rbi.md:207]:
```suggestion
> **The `srb rbi hidden-definitions` command is in
```

File: website/docs/rbi.md (hunk 16 above)
Comment (Morriar) [reply website/docs/rbi.md:131]: Done in 4cf15c320 👍

File: website/docs/rbi.md (hunk 18 above)
Comment (Morriar) [reply website/docs/rbi.md:143]: Done in 7243e6740 👍

File: website/docs/rbi.md (hunk 20 above)
Comment (Morriar) [reply website/docs/rbi.md]: Done in 6f8a4ff4c 👍

File: website/docs/rbi.md (hunk 19 above)
Comment (Morriar) [reply website/docs/rbi.md:165]: Done in f099f437d 👍

File: website/docs/rbi.md [hunk 21]
```diff
@@ -151,42 +155,79 @@ a method's arity, they can also declare a method's input and output types.
...
-version control.
+RBI files in an `rbi/` directory. When they do this, Tapioca will automatically
+merge those definitions in the autogenerated RBIs. In the future, we anticipate
+this to be the preferred way to include RBI files into a project.
+
+## DSL RBIs
+
+Sorbet by itself does not understand DSLs involving meta-programming, such as
+Rails. This means that Sorbet won't know about constants and methods generated
+by `ActiveRecord` or `ActiveSupport`. To solve this, Tapioca can load your
```
[the remark this answers is not in the document]
Comment (Morriar) [reply website/docs/rbi.md:200]: Done in 49f6de5d9 👍

File: website/blog/2022-07-28-srb-tapioca.md (hunk 17 above)
Comment (Morriar) [reply website/blog/2022-07-28-srb-tapioca.md]: Done in 36ae5d915 👍

**Review by jez — APPROVED** [review]
